### ISSUE TITLE
RqLive wrongly assumes the stream has ended when no data is aviable

### ISSUE THREAD

**Issue by laeubi** [yegor256/takes#1300]
`org.takes.rq.RqLive.data(InputStream, Opt<Integer>, boolean)` has the following code

https://github.com/yegor256/takes/blob/20467cf2729566461ad4c7fe7123cf4ea57c51ed/src/main/java/org/takes/rq/RqLive.java#L183-L185

this is wrong, as it indicates that the stream has reached EOF but it simply means the buffer is empty. See also the javadoc here:

> Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without blocking, **which may be 0, or 0 when end of stream is detected**. 

So in case of 0 return value one can simply not tell anything, so at best one can return EMPTY here.

But looking at the code I don't see how this ever is useful in the parsing here, as even if the buffer is empty now, all one can do is to wait for one byte so blocking I/O seems the only option here.

Another pitfall is that here:
https://github.com/yegor256/takes/blob/20467cf2729566461ad4c7fe7123cf4ea57c51ed/src/main/java/org/takes/rq/RqLive.java#L73

it assumes EOF also when the read byte is == 0, even though it should not happen and is an error in the http stream, a zero byte *can* be read from a stream and is not an EOF!

**Comment by piotrkot** [comment]
Thanks @laeubi. That's interesting.
I use `takes` in production and indeed some recent errors I've encountered can come from this code. Do you have an idea how one could write a good test for it? With a fake implementation for InputStream, maybe?

**Comment by piotrkot** [comment]
I think the same mistake, assuming that the `available()` method always returns an accurate estimate of available bytes, occurs in https://github.com/yegor256/takes/blob/2147d98a33aef606f2ef0d793f93a7815e49a798/src/main/java/org/takes/http/BkBasic.java#L88, and https://github.com/yegor256/takes/blob/2147d98a33aef606f2ef0d793f93a7815e49a798/src/main/java/org/takes/servlet/ServletInputStreamOf.java#L37
@laeubi would you agree?

**Comment by piotrkot** [comment]
@yegor256 can I make a PR for it? Do I need to be an assignee first?

**Comment by yegor256** [comment]
@piotrkot sure, please go ahead

**Comment by laeubi** [comment]
> Do you have an idea how one could write a good test for it? With a new implementation for InputStream for which available() method returns always 0, maybe?

Yes this would be a good testcase, in fact that is what can happen as well when using a stream that do not uses any buffers internally.

> I think the same mistake, assuming that the available() method always returns an accurate estimate of available bytes, occurs in 

Yes there are other places as well, the `available()` method should simply never be used anywhere in `takes` code as there is no guarantee it ever will return a value > 0.

So if you like to help writing tests that would be great, there are some more issues with the Http handling but I sadly have not had yet time to further process them due to some other projects having higher precedence and I can usually use some kind of workarounds here.

**Comment by laeubi** [comment]
@piotrkot As mentioned there are other issues as well 

se for example

https://github.com/yegor256/takes/pulls/laeubi

and some of the issues I reported / commented.

> That means RqLive can actually contain multiple requests read from InputStream. It is confusing.

To summarize, currently the keep-alive is not implemented correctly, it should never depend on the availability of bytes and the client need to ensure proper content-length or chuncked encoding or other means are used to find the right bytes. Also without a keep-alive header the server must close the connection otherwise.

If you are interested here I would be happy to give further guidance and will look where changes are needed. Some are tricky to implement with the current architecture though.

**Comment by piotrkot** [comment]
@laeubi I agree there may be many more issues. But we don't want to fix them all in this PR, do we.
Since takes supports HTTP/1.1, connection should not close without keep-alive header, according to https://datatracker.ietf.org/doc/html/rfc7230#section-6.3.
As for "proper content-length or chuncked encoding", I would prefer to resolve them in other PR.

@yegor256 I don't know. How about?
```
interface MultiRequest {
  Iterable<Request> requests();
}

class MultiRqLive implements MultiRequest
```
Then, https://github.com/yegor256/takes/blob/7a351a375c8c1ac8fe2c11a7063622693a87b00d/src/main/java/org/takes/http/BkBasic.java#L98 would iterate over `Requests` instead of the `while(true)` loop.
Does it have sense?

**Comment by laeubi** [comment]
> As for "proper content-length or chuncked encoding", I would prefer to resolve them in other PR.

You cant implement proper keep alive according to HTTP1.1 because that is a *mandatory* for a keep alive session.

Also note that

>  HTTP  implementations _SHOULD_ support persistent connections.

So it is an optional part of the specification, but

> In order to remain persistent, all messages on a connection need to  have a self-defined message length (i.e., one not defined by closure of the connection), as described in [Section 3.3](https://datatracker.ietf.org/doc/html/rfc7230#section-3.3).  A server _MUST_ read the entire request message body or close the connection after sending its response, since otherwise the remaining data on a persistent connection would be misinterpreted as the next request.

So the handling of the headers is actually **mandatory** ...

I'm not sure why one needs a `MultiRequest` that do not sound right, as you correctly have correctly said there is already a loop and the BkBasic must simply not use `input.available()` but properly handle the Http headers to limit access to the stream (what the will return EOF once request is done) then send the response and go on with the next one.

**Comment by piotrkot** [comment]
@laeubi There is one more thing.
According to the specs: "In order to remain persistent, all messages on a connection need to have a self-defined message length". This is clear for POST messages. But what if we want to send GET messages and keep them persistent? Can't we do that?
But then GET messages would need to contain, say, `Content-Length`. Does it have sense?

**Comment by laeubi** [comment]
> According to the specs: "In order to remain persistent, all messages on a connection need to have a self-defined message length". This is clear for POST messages. But what if we want to send GET messages and keep them persistent?

@piotrkot a GET message has an (implicit) self-defined message length as it is always zero.

In general a HTTP Message contains of two parts:

1. the header
2. the payload or message-body

depending on the VERB (e.g. GET, POST, ...) this can have different meanings for the Client and the Server. e.g. a HEAD request will not return any payload neither for the Client nor the server, POST will have one for the Server and possibly for the Client and so on.

The spec says:

> The presence of a message-body in a request is signaled by the inclusion of a Content-Length or Transfer-Encoding header field in the request's message-headers.

So on a persistent connection you need to consider these things:

- the `Keep-Alive` header, both client and server can terminate such connection or ask for keep it alive so `BKBasic` needs to peek into the request/response headers to see if either side has requested a connection close and then close the streams
- the Request itself must limit access to the input stream through `RqLengthAware` so a user can read the stream until EOF but not close it (if there is content (e.g. PUT/POST)) this is something I tried to address here https://github.com/yegor256/takes/pull/1346 ) 
- For request without content it must simply always use an empty stream, I tried to address this here: https://github.com/yegor256/takes/pull/1314

> I was planning to split it into two requests and process sequentially. 
> Now, assume the first request has a Content-Length but with a huge value

Don't try to think about you can "know" how many request or what type of requests are incoming can be know in advance, a usual processing queue for a persistent connection works the following way:

- Assume we have two stream `S_Input` and `S_Output` (that here map to the raw socket streams)
- You read the `request headers` from `S_Input`
- You now know what kind of request it is and if it uses `Content-Length` or `Transfer-Encoding`
- Now you need to either signal an error (e.g. unsupported `Transfer-Encoding`) and close the streams after sending the response or create a new wrapped Stream `W_Input` that delegates to `S_Input` and decodes the chuncks until it founds the end or it uses the content length to read up to Content-Length bytes.
- This is now passed to the handler of the request and let it produce the response headers (send to `S_Output`) and depending on how we are having a result message e.g. in form of a stream (=chuncked) or byte array (=content length) supply it with a wrapped `W_Output` that either forwards or encodes the data.
- If the handler throws an exception, one must still read the full message bytes and  discards them or one might decide to simply close the connection to be safe
- Then if both server+client do not have send a Connection-close, read the next request from the stream

As you can see this **never** requires you to cache anything (except a few bytes) from the request or response, this all can happen using streams.

**Comment by piotrkot** [comment]
@laeubi Thanks for clarification. Still, I have one problem.

We do have a unit test in https://github.com/yegor256/takes/blob/dd71c4f64cca51f359680da004a6439f37d3b050/src/test/java/org/takes/http/BkBasicTest.java#L155 which assumes two requests should be read without explicit `Keep-Alive` header. This is correct according to the specification. And I don't think we should remove this feature.

Now, if I don't know by reading headers that this is the last request in this connection, and I can't trust `available()` method either, what else can I do to tell that we should stop reading? The only thing that comes to me is to try to `read()`. But `read()` method is blocking. Should I add a `setSoTimeout()` with a reasonable duration in the `Socket` and continue when `SocketTimeoutException` is thrown? It doesn't seem right.

**Comment by laeubi** [comment]
@piotrkot two things:

1. The server is in control! So if the server things it do not likes to wait any longer, it is allowed to close the connection. One can inform the client about this e.g. with `Keep-Alive: timeout=15` seconds (what at best is configurable somehow).
2. If the client no longer wants to use the connection it should send `Connection: close` and drop the connection after getting the response from the server (that then also should respond with `Connection: close` and drop the connection)

For the first, one can use different techniques, one might `setSoTimeout` or a watchdog that closes the socket. In all these cases a `SocketException` will break out the read. Of course non-blocking I/O would be better here, but we should not complicate it too much. In any case the code needs to check if a `Connection: close` was send by the client and then not try to perform another read.

For the second, such test should behave nice and inform the server after the second request that it do not plan to further use the connection so it does not hang forever. This then also can include a timeout to assert the server is really closing the connection as expected.
